homeassistant/components/tplink_omada/switch.py

```diff
@@ -109,3 +177,72 @@ def _handle_coordinator_update(self) -> None:
         """Handle updated data from the coordinator."""
         self.port_details = self.coordinator.data[self.port_id]
         self._refresh_state()
+
+
+class OmadaGatewayPortSwitchEntity(OmadaDeviceEntity[OmadaGateway], SwitchEntity):
+    """Generic toggle switch on a Gateway entity."""
+
+    _attr_has_entity_name = True
+    _port_details: OmadaGatewayPortStatus | None = None
+    entity_description: GatewayPortSwitchEntityDescription
+
+    def __init__(
+        self,
+        coordinator: OmadaGatewayCoordinator,
+        device: OmadaGateway,
+        port_number: int,
+        entity_description: GatewayPortSwitchEntityDescription,
+    ) -> None:
+        """Initialize the toggle switch."""
+        super().__init__(coordinator, device)
+        self.entity_description = entity_description
+        self._port_number = port_number
+        self._attr_unique_id = f"{device.mac}_{port_number}_{entity_description.key}"
+        self._attr_translation_placeholders = {"port_name": f"{port_number}"}
+
+    async def async_added_to_hass(self) -> None:
+        """When entity is added to hass."""
+        await super().async_added_to_hass()
+        self._do_update()
+
+    async def _async_turn_on_off(self, enable: bool) -> None:
+        if self._port_details:
```

Comment:
Thanks. I will look at improving this in the next PR. 

There is a window of possibility if a port is re-configured on the router it could stop being a WAN port, for example, which could result in `_port_details` becoming `None`. I'm not really sure how that sort of dynamic change is supposed to be handled - whether it should just go unavailable until the next HA restart, or whether I should put in logic to add/remove entities when the hub configuration changes. At the moment, the integration only adds entities at startup, so dynamic changes to the network won't be reflected.

Reply:
If we can detect changes we can add/remove entities.